core/modules/taxonomy/taxonomy.module

```diff
@@ -1482,11 +1489,118 @@ function taxonomy_field_formatter_view($entity_type, $entity, $field, $instance,
+        list($id) = entity_extract_ids($entity_type, $entity);
+        $recursion_queue[$id] = $id;
+      }
+
+      // Check the recursion queue to determine which terms should be fully
+      // displayed, and which terms will only be displayed as a title.
+      $terms_display = array();
+      foreach ($items as $delta => $item) {
+        if (!isset($recursion_queue[$item['tid']])) {
+          $terms_display[$item['tid']] = $item['taxonomy_term'];
+        }
+      }
+
+      // Load and build the fully displayed terms.
+      if ($terms_display) {
+        foreach ($terms_display as $tid => $term) {
+          $terms_display[$tid]->referencing_entity = $entity;
+          $terms_display[$tid]->referencing_field = $field['field_name'];
+        }
+        $terms_built = taxonomy_term_view_multiple($terms_display, $settings['taxonomy_term_reference_view_mode']);
+      }
+
+      // Assemble the render array.
+      foreach ($items as $delta => $item) {
+        if (isset($terms_display[$item['tid']])) {
+          $element[$delta] = $terms_built['taxonomy_terms'][$item['tid']];
+        }
+        else {
+          $term = $item['taxonomy_term'];
+          $label = entity_label('taxonomy_term', $term);
+          $uri = entity_uri('taxonomy_term', $term);
+          $element[$delta] = array(
+            '#type' => 'link',
+            '#title' => $label,
+            '#href' => $uri['path'],
+            '#options' => $uri['options'],
+          );
+        }
+      }
+
+      break;
   }
 
   return $element;
 }
 
+/**
+ * Implements hook_field_formatter_settings_form().
+ */
+function taxonomy_field_formatter_settings_form($field, $instance, $view_mode, $form, &$form_state) {
+  $display = $instance['display'][$view_mode];
+  $settings = $display['settings'];
+
+  $element = array();
+
+  switch ($display['type']) {
+    case 'taxonomy_term_reference_term':
+      $entity_info = entity_get_info('taxonomy_term');
+      $modes = $entity_info['view modes'];
+      $options = array();
+      foreach ($modes as $name => $mode) {
+        $options[$name] = $mode['label'];
+      }
+      $element['taxonomy_term_reference_view_mode'] = array(
+        '#title' => t('View mode'),
+        '#type' => 'select',
+        '#options' => $options,
+        '#default_value' => $settings['taxonomy_term_reference_view_mode'],
+        // Never empty, so no #empty_option
+      );
+      break;
+  }
+  return $element;
+}
+
+/**
+ * Implements hook_field_formatter_settings_summary().
+ */
+function taxonomy_field_formatter_settings_summary($field, $instance, $view_mode) {
+  $display = $instance['display'][$view_mode];
+  $settings = $display['settings'];
+  $summary = array();
+
+  switch ($display['type']) {
+    case 'taxonomy_term_reference_term':
+      $entity_info = entity_get_info('taxonomy_term');
+      $modes = $entity_info['view modes'];
+      $mode = $modes[$settings['taxonomy_term_reference_view_mode']]['label'];
```

Comment:
Doesn't seem like "Default" comes up in the options, so it might confuse people if that appears in the summary.

![screenshot from 2018-09-12 00-34-49](https://user-images.githubusercontent.com/1071483/45402472-cd57e400-b623-11e8-9352-595a72179b5d.png)

Reply:
I'm going to leave it as "full" for now but perhaps we can discuss whether we want to change "Taxonomy term page" to something else.